components/asset-store-controller-manager/pkg/controller/asset/asset_controller.go

```diff
@@ -136,96 +153,101 @@ func (r *ReconcileAsset) Reconcile(request reconcile.Request) (reconcile.Result,
 		return reconcile.Result{}, err
 	}
 
-	switch {
-	case !r.handler.ShouldReconcile(instance, instance.Status.CommonAssetStatus, time.Now(), time.Minute):
-		return reconcile.Result{}, nil
-	case r.handler.IsOnDelete(instance):
-		return r.onDelete(ctx, instance)
-	case r.handler.IsOnAddOrUpdate(instance, instance.Status.CommonAssetStatus):
-		return r.onAddOrUpdate(ctx, instance)
-	case r.handler.IsOnPending(instance.Status.CommonAssetStatus):
-		return r.onPending(ctx, instance)
-	case r.handler.IsOnReady(instance.Status.CommonAssetStatus):
-		return r.onReady(ctx, instance)
-	case r.handler.IsOnFailed(instance.Status.CommonAssetStatus):
-		return r.onFailed(ctx, instance)
-	}
-
-	return reconcile.Result{}, nil
-}
-
-func (r *ReconcileAsset) onDelete(ctx context.Context, instance *assetstorev1alpha2.Asset) (reconcile.Result, error) {
-	if !r.finalizer.IsDefinedIn(instance) {
-		return reconcile.Result{}, nil
+	assetLogger := log.WithValues("kind", instance.GetObjectKind().GroupVersionKind().Kind, "name", instance.GetName(), "namespace", instance.GetNamespace())
+	commonHandler := asset.New(assetLogger, r.recorder, r.store, r.loader, r.findBucketFnc, r.validator, r.mutator, r.relistInterval)
+	commonStatus, err := commonHandler.Handle(ctx, time.Now(), instance, instance.Spec.CommonAssetSpec, instance.Status.CommonAssetStatus)
+	if updateErr := r.updateStatus(ctx, request.NamespacedName, commonStatus); updateErr != nil {
+		finalErr := updateErr
+		if err != nil {
+			finalErr = errors.Wrapf(err, "along with update error %s", updateErr.Error())
+		}
+		return reconcile.Result{}, finalErr
 	}
-
-	err := r.handler.OnDelete(ctx, instance, instance.Spec.CommonAssetSpec)
 	if err != nil {
 		return reconcile.Result{}, err
 	}
 
-	r.finalizer.DeleteFrom(instance)
-	if err := r.Update(ctx, instance); err != nil {
-		return reconcile.Result{}, errors.Wrap(err, "while updating instance")
+	if err := r.removeFinalizer(ctx, request.NamespacedName); err != nil {
```

Comment:
Hmm, shouldn't this function be put to the common handler? We could pass an optional function `onDelete` or something like that

Reply:
Hm... right now I don't see a nice way to move it to a common handler. If you have some ideas, please share 😄